octobot_trading/exchanges/parser/util.py

```diff
@@ -0,0 +1,424 @@
+import datetime
+import decimal
+from octobot_commons import timestamp_util
+from octobot_trading import constants as trading_constants
+from octobot_trading.enums import (
+    ExchangePositionCCXTColumns,
+)
+
+
+class Parser:
+    """
+    overwrite Parser class methods if necessary
+    always/only include bulletproof custom code into the parser to improve generic support
+    """
+
+    def __init__(self, exchange):
+        self.PARSER_TITLE = ""  # example for position parser:  "position"
+        self.REQUIRED_FIELDS = []  # keys that need to be in the formatted record
+        self.NON_EMPTY_FIELDS = []  # keys that can't be empty in the formatted record
+
+        self.debugging_report_dict = {}
+        self.debugging_report_str = ""
+        self.missing_attributes_str = ""
+
+        self.raw_records = {}
+        self.exchange = exchange  # todo replace with logger
+        self.raw_record = {}
+        self.formatted_record = {}
+        self.formatted_records = []
+        self.multiple_records = False
+
+    def _ensure_details_completeness(self):
+        # ensure all REQUIRED_FIELDS are present and all NON_EMPTY_FIELDS are not empty
+
+        return all(
+            key in self.formatted_record for key in self.REQUIRED_FIELDS
+        ) and all(self.formatted_record[key] for key in self.NON_EMPTY_FIELDS)
+
+    def _create_debugging_report_for_record(self):
+        if not self.multiple_records:
+            self._create_debugging_report()
+
+    def _create_debugging_report(self):
+        if self.debugging_report_dict:
+            # filter duplicated errors
+            debugging_report_str = ""
+            missing_attributes_str = ""
+            for (
+                attribute_name,
+                error_messages_list,
+            ) in self.debugging_report_dict.items():
+                for index in range(len(error_messages_list)):
+                    if not error_messages_list[index] in error_messages_list[:index]:
+                        debugging_report_str += error_messages_list[index]
+                missing_attributes_str += f"{attribute_name}, "
+
+            self.exchange.logger.error(
+                self.debugging_report_template(
```

Comment:
I think this debugging report is a good idea but shouldn't managed only in parsers, it should be an octobot common thing

Reply:
I moved it to commons https://github.com/Drakkar-Software/OctoBot-Commons/pull/293